Utils/Subject.h

```diff
@@ -1,104 +1,270 @@
 #pragma once
 
-#include "Observer.h"
-
 #include <vector>
 #include <functional>
+#include <memory>
+
+#include "ReflectionUtils.h"
 
 namespace leap
 {
 	class Subject
 	{
-	public:
-		Subject() = default;
-		~Subject()
+	private:
+		struct MemberFunctionWrapper
 		{
-			for (auto& observer : m_Observers)
+			virtual ~MemberFunctionWrapper() = default;
+			virtual void Call() const = 0;
+			virtual unsigned int GetId() const = 0;
+			virtual bool IsConst() const = 0;
+		};
+
+		template<typename T>
+		struct ImplMemberFunctionWrapper final : public MemberFunctionWrapper
+		{
+			ImplMemberFunctionWrapper(T* _pThis, void(T::* _pFunction)())
+				: pThis{ _pThis }, pFunction{ _pFunction }, id{ ReflectionUtils::GenerateTypenameHash<T>() }
 			{
-				observer->OnSubjectDestroy();
 			}
-		}
+			virtual ~ImplMemberFunctionWrapper() = default;
 
-		void AddListener(Observer* observer)
+			unsigned int id{};
```

Comment:
All member variables publicly exposed.
I know its internal but you also made a GetId() function for a publicly exposed id member var.